src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java

```diff
@@ -2,37 +2,64 @@
 
 import com.iridium.iridiumcore.utils.StringUtils;
 import com.iridium.iridiumskyblock.IridiumSkyblock;
-import com.iridium.iridiumskyblock.api.IridiumSkyblockAPI;
+import com.iridium.iridiumskyblock.api.IslandEnterEvent;
+import com.iridium.iridiumskyblock.api.IslandLeaveEvent;
 import com.iridium.iridiumskyblock.database.Island;
 import com.iridium.iridiumskyblock.database.IslandBooster;
 import com.iridium.iridiumskyblock.database.User;
+import org.bukkit.Bukkit;
 import org.bukkit.GameMode;
 import org.bukkit.entity.Player;
 import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
 import org.bukkit.event.Listener;
 import org.bukkit.event.player.PlayerMoveEvent;
 
 import java.util.Optional;
 
 public class PlayerMoveListener implements Listener {
 
-    @EventHandler(ignoreCancelled = true)
+    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
     public void onPlayerMove(PlayerMoveEvent event) {
         Player player = event.getPlayer();
         User user = IridiumSkyblock.getInstance().getUserManager().getUser(player);
-        if (event.getTo().getBlockX() != event.getFrom().getBlockX() || event.getTo().getBlockZ() != event.getFrom().getBlockZ()) {
+        if (event.getTo() != null && (event.getTo().getBlockX() != event.getFrom().getBlockX() || event.getTo().getBlockZ() != event.getFrom().getBlockZ())) {
             if (user.getTeleportingTask() != null) {
                 user.getTeleportingTask().cancel();
+
                 user.setTeleportingTask(null);
                 player.sendMessage(StringUtils.color(IridiumSkyblock.getInstance().getMessages().teleportCanceled
                         .replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix))
                 );
             }
+            Optional<Island> fromIsland = IridiumSkyblock.getInstance().getIslandManager().getIslandViaLocation(event.getFrom());
```

Comment:
I dont like this, will be verry laggy. I was thinking about implementing some sort of cache for this, when we get the island a player is on since like 99% of the time that island wont have changed from the last time we called it